content/blog/2023-12-11-bloke/index.md

```diff
@@ -0,0 +1,372 @@
+
+For simplicity, we only consider loop-free Bril programs with integer return values and only primitive operations. Adding support for more Bril operations such as function calls is future work.
+
+![](https://cdn.mathpix.com/cropped/2023_12_10_5f34f3842a7018d69e08g-04.jpg?height=421&width=578&top_left_y=930&top_left_x=354)
+
+Figure 2: BLOKE optimization of clobber
+
+Figure 2 shows the optimization BLOKE may produce when Figure 1 is the input. We successfully find that the last three lines of Figure 1 are the only necessary lines for an optimized equivalent program.
+
+## Architecture
+
+Figure 3 shows high-level architecture of BLOKE. The user would call the optimizer library which samples using STOKE's random walk method. The sample library can call the Brilirs interpreter to run test cases, or call the bril\_equivalence library to prove that a program in the walk is equivalent to the input program. The equivalence library would lift the Bril programs into $\mathrm{Z} 3$ formulas using the bril2z3 library and call $Z 3$ to prove equivalence or find a counterexample.
+
+![](https://cdn.mathpix.com/cropped/2023_12_10_5f34f3842a7018d69e08g-05.jpg?height=352&width=1482&top_left_y=239&top_left_x=316)
+
+Figure 3: High level BLOKE architecture
+
+## Correctness Cost
+
+Since we have integer outputs, we can use the usual Euclidean metric to define the validation cost similar to STOKE:
+
+$$
+\text{val}(x)=\sum\_{(i, o) \in \tau}|x(i)-o|+\text{err}(x, i)
+$$
+
+Currently we set $\text{err}(x, i)=2$ for all inputs. Varying the error cost based on the error type is future work.
+
+## SMT Bril Equivalence checking
+
+For the verification cost, we adopted Equation 1 and wrote two libraries, bril2z 3 and bril\_equivalence.
+
+bril2z3 allows conversion of Bril programs to $\mathrm{Z} 3$ formulas. Since $\mathrm{Z} 3$ formulas do not have a concept of variable assignment, we need to morph the Bril program to a form where each variable used in the program has one and exactly one definition.
+
+![](https://cdn.mathpix.com/cropped/2023_12_10_5f34f3842a7018d69e08g-04.jpg?height=583&width=685&top_left_y=1240&top_left_x=1124)
+
+Figure 4: SSA form of clobber
+
+This is called the static single-assignment (SSA) form of a program, and there are existing algorithms to convert any program written in an intermediate representation language like Bril to SSA form [7].
+
+![](https://cdn.mathpix.com/cropped/2023_12_10_5f34f3842a7018d69e08g-05.jpg?height=416&width=678&top_left_y=694&top_left_x=304)
+
+Figure 5: Z3 formula for clobber
+
+With the SSA form of the loop-free program, we can generate the $\mathrm{Z} 3$ formula of the program. We encode all control flow paths of the loop-free program into Z3 formulas and encode the return value of the program as a special variable name with a datatype that supports all Bril primitive types. Figure 5 is an example Z3 formula of clobber.
```

Comment:
I think it would be more accurate to say "SMT formula" than "Z3" formula, since this is the standard SMTLIB format.

Reply:
Should we replace all instances of Z3 with SMT, and mention that we used Z3 to generate the SMT formulas and prove equivalence?